### PR TITLE
Implement infinite scrolling / autoplay carousel.

### PR DESCRIPTION
**Pull request by DylanPiercey**
## Description
Adds `autoplay` and related attributes for an autoplaying carousel.

In this PR there was also some issues resolved around canceling native scroll transitions and a rounding error when determining the slide sizes.

# Notes
* Currently the native scrolling is disabled for autoplay carousels (can still navigate via controls).
* New events and autoplay related attributes are mentioned in the updated README.

## References
Implements #216

## Screenshots
image

### CODE REVIEW

**Comment by DylanPiercey** [comment]
Note I plan on adding more tests with this PR, will do so tomorrow morning.

File: src/components/ebay-carousel/README.md [hunk 1]
```diff
@@ -20,15 +20,24 @@ Name | Type | Stateful | Description
...
+`items-per-slide` | String | No | automatically fit a number of items for each carousel slide and enable slide controls
+`gap` | String | No | override the margin between carousel items in pixels (default: "16")
 
 ### Additional Attributes for when items-per-slide is set.
 Name | Type | Stateful | Description
 --- | --- | --- | ---
 `accessibility-status` | String | No | status text (default: "Showing Slide {currentSlide} of {totalSlides} - Carousel")
 `accessibility-current` | String | No | pagination current slide text (default: "Current Slide {currentSlide} - Carousel")
 `accessibility-other` | String | No | pagination other slide text (default: "Slide {slide} - Carousel")
+`autoplay` | Boolean | No | adds autoplay controls to the carousel and automatically goes to the next slide every 4 seconds
```
Comment (yomed) [inline src/components/ebay-carousel/README.md]: Was there any need for config for the `interval` timing?

File: src/components/ebay-carousel/examples/10-autoplay/template.marko [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<style>
```
Comment (yomed) [inline src/components/ebay-carousel/examples/10-autoplay/template.marko]: Regarding the numbers in the folders -- did you prefer this to sorting?
https://github.com/eBay/ebayui-core/blob/1343c7b2c895669d812694fce693fbc69a7eb842/demo/index.js#L51

File: src/components/ebay-carousel/examples/10-autoplay/template.marko [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    .demo7-card {
```
Comment (yomed) [inline src/components/ebay-carousel/examples/10-autoplay/template.marko]: The demo class setup isn't great here to begin with, but you're probably redeclaring the `demo7-...` classes from above.

File: src/components/ebay-carousel/index.js [hunk 4]
```diff
@@ -128,21 +151,26 @@ function onRender() {
...
-            const offset = getOffset(state);
+            const offset = hasOverride ? offsetOverride : getOffset(state);
+            // Listen for future scroll events and snap to the closest point.
+            this.cancelScrollHandler = onScrollEnd(listEl, handleScrollEnd.bind(this));
             // Animate to the new scrolling position and emit update events afterward.
             this.cancelScrollTransition = scrollTransition(listEl, offset, this.emitUpdate);
         }
 
+        if (autoplay && !paused) {
+            this.autoPlayTimeout = setTimeout(() => this.nextEl.click(), 4000);
```
Comment (yomed) [inline src/components/ebay-carousel/index.js]: (minor) `autoplay` is used as 1 word elsewhere

File: src/components/ebay-carousel/style-touch.less [hunk 5]
```diff
@@ -19,6 +19,12 @@
         }
     }
 
+    &__autoplay {
+        .carousel__list {
```
Comment (yomed) [inline src/components/ebay-carousel/style-touch.less]: (minor) Can leave out the nesting here

File: src/components/ebay-carousel/style.less [hunk 6]
```diff
@@ -67,6 +69,35 @@
...
+        padding: 0;
+        background-color: rgba(0,0,0,0.45);
+        height: 40px;
+        width: 40px;
+        border-radius: 50%;
+        text-align: center;
+        color: @ds6-color-g201-gray;
+        box-sizing: border-box;
+
+        &&& svg {
```
Comment (yomed) [inline src/components/ebay-carousel/style.less]: I think you were aiming for 3 play and 3 pause, but it ends up like 
```css
.carousel__play.carousel__play.carousel__play svg, 
.carousel__play.carousel__play.carousel__pause svg, 
.carousel__play.carousel__pause.carousel__play svg, 
.carousel__play.carousel__pause.carousel__pause svg, 
.carousel__pause.carousel__play.carousel__play svg, 
.carousel__pause.carousel__play.carousel__pause svg, 
.carousel__pause.carousel__pause.carousel__play svg,
.carousel__pause.carousel__pause.carousel__pause svg
```

File: src/components/ebay-carousel/style.less [hunk 7]
```diff
@@ -67,6 +69,35 @@
...
     }
 
+    &__play, &__pause {
+        z-index: 2;
+        bottom: 16px; // accounts for padding.
+        right: 0;
+        margin-right: 19px;
+        margin-bottom: 19px;
+        position: absolute;
+        background: 0;
```
Comment (yomed) [inline src/components/ebay-carousel/style.less]: Is this needed with `background-color` below?

File: src/components/ebay-carousel/README.md (hunk 1 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/README.md]: I based the interval on the homepage interval. However it makes it easier to test to make this configurable so I updated the `autoplay` attribute to be either a Boolean or Number. If a number is set that becomes the interval.

File: src/components/ebay-carousel/examples/10-autoplay/template.marko (hunk 2 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/examples/10-autoplay/template.marko]: The ui-marketplace uses folder sorting order which is why I made this update.

File: src/components/ebay-carousel/index.js (hunk 4 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/index.js]: Will switch to all being `autoplay`. 👍

File: src/components/ebay-carousel/style-touch.less (hunk 5 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/style-touch.less]: I need the nesting here though since `carousel__list` is a child of `carousel__autoplay` (which is on the root element).

File: src/components/ebay-carousel/style.less (hunk 7 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/style.less]: This was just c&p'd from homepage. Will see if we actually need it.

File: src/components/ebay-carousel/style.less (hunk 6 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/style.less]: The more `&`'s the more frustrated I am with css. (Will try to clean this up)

File: src/components/ebay-carousel/examples/10-autoplay/template.marko (hunk 3 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/examples/10-autoplay/template.marko]: Good catch thanks!

File: src/components/ebay-carousel/README.md (hunk 1 above)
Comment (senthilp) [reply src/components/ebay-carousel/README.md]: We should not make the `interval` configurable, as that can lead to abuse. If 4 second is what was standardized then we will stick with it.

File: marko.json [hunk 8]
```diff
@@ -37,12 +37,17 @@
         "@index": "string",
         "@type": "string",
         "@slide": "string",
+        "@autoplay": "boolean",
+        "@paused": "boolean",
```
Comment (ianmcburnie) [inline marko.json]: Setting a `paused` property feels a little odd to me. I would have thought we'd use `play()` and `pause()` methods similar to the HTML video tag. I would imagine a `paused` attribute would be read-only (again, like the video tag) if we are trying to be HTML-like. @senthilp thoughts?

Comment (DylanPiercey) [reply marko.json]: I took inspirited from the video tags `muted` property. We could expose methods for this as well if we want though.

Comment (DylanPiercey) [reply marko.json]: Also I am exposing the stateful `paused` property on the element (which the video tag also does). However the video element exposes 10^10 ways of configuring it's state so 🤷‍♂️. Previously I think we decided against having methods for boolean toggles like this.

Comment (senthilp) [reply marko.json]: Stateful `paused` makes sense, very similar to `video` tag's `paused`. This is sort of an in-between situation where there is no right or wrong.

Comment (ianmcburnie) [reply marko.json]: @senthilp Video tag's `paused` is read-only.

Comment (ianmcburnie) [reply marko.json]: ... and of course this is a carousel, not video, I know I know ;-) But trying to find some common ground.

Comment (ianmcburnie) [reply marko.json]: I think `muted` is just an attribute for initial render, and not stateful:

`muted - Whether to mute the media resource by default`

https://www.w3.org/TR/html51/semantics-embedded-content.html#the-video-element

Comment (DylanPiercey) [reply marko.json]: I am okay with either option. Making `paused` not stateful and switching to `play()` and `pause()` methods is also fine, could make the docs slightly more confusing though.

Comment (ianmcburnie) [reply marko.json]: But `paused` will make no sense on 99% of our carousels, right? Only the ones with autoplay.

Comment (DylanPiercey) [reply marko.json]: That is correct.

Comment (ianmcburnie) [reply marko.json]: I would remove `paused` in order to fix this anomaly.

Comment (ianmcburnie) [reply marko.json]: Actually, let me think on that last comment a bit more ;-)

Comment (ianmcburnie) [reply marko.json]: My conclusion is... that they should have named it `automute` rather than `muted`. Lol.

I can see the attraction and simplicity of just a single stateful property to control playback (especially as we have the power to do that in our Marko layer on top of HTML), but if we are striving to be like the HTML API then we should keep `paused` as a readonly property, and add play() and pause() methods. This will be most familiar for users of the video tag. Not saying it's better, just more familiar ;-)

In all honesty I don't mind either way, but it feels like as we've been strict with the API up to now, we may as well continue on this path. I think after v1 we should asses all of the components again (properties vs methods) as a whole and see if it's still the right approach.

Comment (DylanPiercey) [reply marko.json]: Are we fine with allowing `paused` to be set programmatically as attributes when using marko but just removing the property? Otherwise we are stuck with changing the state imperatively which is lame 😕.

File: src/components/ebay-carousel/README.md (hunk 1 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/README.md]: @senthilp at first I agreed with you on this (which is why my original implementation hardcoded the 4s interval) however after giving it some thought I think this might not be the right approach in this case. The interval for the carousel needs to be finely tuned to the content that is in the carousel such that the user has enough time to properly consume that content. I imagine that could change enough to warrant having it be configurable.

File: marko.json (hunk 8 above)
Comment (senthilp) [reply marko.json]: @ianmcburnie I tried the `muted` attribute on YouTube `video` tag and it is stateful. I was able to mute and unmute by doing `$0.muted = true` and `$0.muted = false` 😄 .

File: src/components/ebay-carousel/README.md (hunk 1 above)
Comment (senthilp) [reply src/components/ebay-carousel/README.md]: @DylanPiercey Agree.

File: marko.json (hunk 8 above)
Comment (ianmcburnie) [reply marko.json]: Hmm, that didn't work for me in Safari. Let me try again...

Comment (ianmcburnie) [reply marko.json]: Adding and removing the attribute is not stateful though. Is that what you meant? Only setting it programmatically works for me (i.e. $0.muted = true and $0.muted = false).

Comment (senthilp) [reply marko.json]: Yes, adding/removing does not work. Only setting it programmatically works.

File: src/components/ebay-carousel/index.js [hunk 9]
```diff
@@ -205,12 +280,30 @@ function handleDotClick(originalEvent, target) {
...
+/**
+ * Toggles the play state of an autoplay carousel.
+ *
+ * @param {MouseEvent} originalEvent
+ */
+function togglePlay(originalEvent) {
+    const { state: { config, paused } } = this;
+    config.preserveItems = true;
+    this.setState('paused', !paused);
+    if (paused) this.nextEl.click();
```
Comment (yomed) [inline src/components/ebay-carousel/index.js]: Did you intend for the `carousel-next` event to be called by this (and the other time .click() is used)? I think we only wanted to call those events from human interaction, so that they can be used for tracking purposes.

File: src/components/ebay-carousel/style-touch.less (hunk 5 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/style-touch.less]: I misunderstood, added a commit to remove the unneeded nesting. Thanks!

File: src/components/ebay-icon/symbols/pause.marko [hunk 10]
```diff
@@ -0,0 +1 @@
+<symbol id="icon-pause" viewBox="0 0 16 16"><g fill="currentColor" fill-rule="evenodd"><path d="M0 0h6v16H0zM10 0h6v16h-6z"></path></g></symbol>
```
Comment (yomed) [inline src/components/ebay-icon/symbols/pause.marko]: Should these be added to the `ebay-icon/examples`?

File: src/components/ebay-carousel/index.js [hunk 11]
```diff
@@ -181,12 +209,59 @@ function emitUpdate() {
...
+    let offsetOverride;
+
     config.preserveItems = true;
+    this.isMoving = true;
+
+    // When we are in autoplay mode we overshoot the desired index to land on a clone
+    // of one of the ends. Then after the transition is over we update to the proper position.
+    if (autoplayInterval) {
+        const LEFT = -1;
+        const RIGHT = 1;
```
[the remark this answers is not in the document]
Comment (DylanPiercey) [reply src/components/ebay-carousel/index.js]: Good call, will do.

File: src/components/ebay-icon/symbols/pause.marko (hunk 10 above)
Comment (DylanPiercey) [reply src/components/ebay-icon/symbols/pause.marko]: I'm not sure. They are not in skin yet so maybe not?

Comment (yomed) [reply src/components/ebay-icon/symbols/pause.marko]: Ah ok, let's leave them out then.

File: src/components/ebay-carousel/index.js (hunk 9 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/index.js]: In the code above it is triggered by clicking the play button and so I think it makes sense, for the actual autoplay part I am not sure, should we have a separate event? Should it not emit any special events? cc @ianmcburnie @senthilp?

File: marko.json (hunk 8 above)
Comment (yomed) [reply marko.json]: I think `paused` is fine as a property since it has a direct (and useful) mapping to the state. Previously we discussed using methods for either 1) dealing with non-primitive data types, or 2) acting on the component without a clear mapping to state.

File: src/components/ebay-carousel/index.js (hunk 9 above)
Comment (ianmcburnie) [reply src/components/ebay-carousel/index.js]: I would prefer:

```js
if (paused) {
   this.gotoNextSlide(); // (or watever our method is called to move to next slide)
}
```

 i.e. call our public or internal API rather than simulate click.

**Comment by yomed** [comment]
@DylanPiercey Not related to specific code, but I think you may need to watch out for autoplaying while resizing the browser.

File: src/components/ebay-carousel/index.js (hunk 9 above)
Comment (DylanPiercey) [reply src/components/ebay-carousel/index.js]: Yeah I agree. Okay I'll make this change.

**Comment by DylanPiercey** [comment]
I think I've addressed everything, let me know if I've missed anything 😄.

**Comment by seangates** [comment]
@DylanPiercey Please fix the conflicts, then I'll look once more, and barring any crazy stuff I see, I'll approve and merge to get the pre-release out.

**Comment by seangates** [comment]
@DylanPiercey Sweet. Looking now ...

**Comment by DylanPiercey** [comment]
Let me try to fix the tests in the ci

File: src/components/ebay-carousel/index.js [hunk 12]
```diff
@@ -80,7 +80,8 @@ function getTemplateData(state) {
 
         // Account for users providing a style string or object for each item.
         if (typeof style === 'string') {
-            item.style = `${style};flex-basis:${itemWidth};margin-right:${marginRight};transform:${transform}`;
+            item.style = `${style};flex-basis:${itemWidth};margin-right:${marginRight}`;
+            if (transform) item.style += `transform:${transform}`;
```
Comment (seangates) [inline src/components/ebay-carousel/index.js]: This passed linting?

Comment (DylanPiercey) [reply src/components/ebay-carousel/index.js]: 👍
